src/frontend-samples/read-settings-sample/ExternalSettingsUI.tsx

```diff
@@ -0,0 +1,169 @@
+import { IModelQuery } from "@bentley/imodelhub-client";
+import { ChangeEvent } from "react";
+import { SettingsResult, SettingsStatus } from "@bentley/product-settings-client";
+
+/** React props */
+interface ExternalSettingsProps {
+  iModelName: string;
+  iModelSelector: React.ReactNode;
+}
+
+/** React state */
+export interface ExternalSettingsState {
+  settingKey?: string;
+  settingValue?: string;
+  settingResult?: SettingsResult;
+  imodel?: IModelConnection;
+  projectId?: string;
+  imodelId?: string;
+  saveInProgress: boolean;
+  status?: string;
+}
+
+const namespace = 'showcase';
+
+/** A React component that renders the UI specific for this sample */
+export default class ExternalSettingsUI extends React.Component<ExternalSettingsProps, ExternalSettingsState> {
+
+  /** Creates an Sample instance */
+  constructor(props?: any, context?: any) {
+    super(props, context);
+    this.state = {
+      saveInProgress: false
+    };
+    this.handleSettingsChange = this.handleSettingsChange.bind(this);
+    this.handleSettingsValueChange = this.handleSettingsValueChange.bind(this);
+    this.saveSettings = this.saveSettings.bind(this);
+  }
+
+  private onIModelReady = (imodel: IModelConnection) => {
+    this.setState({ imodel });
+    AuthorizedFrontendRequestContext.create().then(async context => {
+      const imodelInfo = await this.getIModelInfo(this.props.iModelName, context);
+      this.setState({ projectId: imodelInfo.projectId, imodelId: imodelInfo.imodelId });
+    });
+  }
+
+  // This method serves to query projectId of project where iModel is stored in this example, 
+  // however in real application you might have it upfront
+  private async getIModelInfo(iModelName: string, context: AuthorizedFrontendRequestContext): Promise<{ projectId: string, imodelId: string }> {
+    // In testdrive the projectName matches iModelName.  That's not true in general.
+    const projectName = iModelName;
+    const connectClient = new ContextRegistryClient();
+    let project: Project;
+
+    try {
+      project = await connectClient.getProject(context, { $filter: `Name+eq+'${iModelName}'` });
+    } catch (e) {
+      throw new Error(`Project with name "${projectName}" does not exist`);
+    }
+
+    const imodelQuery = new IModelQuery();
+    imodelQuery.byName(iModelName);
+    const imodels = await IModelApp.iModelClient.iModels.get(context, project.wsgId, imodelQuery);
+    if (imodels.length === 0)
+      throw new Error(`iModel with name "${iModelName}" does not exist in project "${projectName}"`);
+
+    return { projectId: project.wsgId, imodelId: imodels[0].wsgId };
+  }
+
+  // Handler to read external settings when name in dropdown changes
+  handleSettingsChange(event: ChangeEvent<HTMLSelectElement>) {
+    const settingKey = event.target.value;
+    this.setState({ settingKey: settingKey });
+
+    AuthorizedFrontendRequestContext.create().then(async context => {
```

Comment:
The code that calls getSetting should be triggered in the componentDidUpdate method.  Also, the implementation should be in a static method in the App.tsx file.

Reply:
Made static and moved to app.tsx